Content.Server/GameTicking/Rules/GameRuleSystem.Utility.cs

```diff
@@ -0,0 +1,138 @@
+﻿using System.Diagnostics.CodeAnalysis;
+using Content.Server.GameTicking.Rules.Components;
+using Content.Server.Station.Components;
+using Robust.Shared.Collections;
+using Robust.Shared.Map;
+using Robust.Shared.Map.Components;
+using Robust.Shared.Random;
+
+namespace Content.Server.GameTicking.Rules;
+
+public abstract partial class GameRuleSystem<T> where T: IComponent
+{
+    protected EntityQueryEnumerator<ActiveGameRuleComponent, T, GameRuleComponent> QueryActiveRules()
+    {
+        return EntityQueryEnumerator<ActiveGameRuleComponent, T, GameRuleComponent>();
+    }
+
+    protected bool TryRoundStartAttempt(RoundStartAttemptEvent ev, string localizedPresetName)
+    {
+        var query = EntityQueryEnumerator<ActiveGameRuleComponent, T, GameRuleComponent>();
+        while (query.MoveNext(out _, out _, out _, out var gameRule))
+        {
+            var minPlayers = gameRule.MinPlayers;
+            if (!ev.Forced && ev.Players.Length < minPlayers)
+            {
+                ChatManager.SendAdminAnnouncement(Loc.GetString("preset-not-enough-ready-players",
+                    ("readyPlayersCount", ev.Players.Length), ("minimumPlayers", minPlayers),
+                    ("presetName", localizedPresetName)));
+                ev.Cancel();
+                continue;
+            }
+
+            if (ev.Players.Length == 0)
+            {
+                ChatManager.DispatchServerAnnouncement(Loc.GetString("preset-no-one-ready"));
+                ev.Cancel();
+            }
+        }
+
+        return !ev.Cancelled;
+    }
+
+    /// <summary>
+    ///     Utility function for finding a random event-eligible station entity
+    /// </summary>
+    protected bool TryGetRandomStation([NotNullWhen(true)] out EntityUid? station, Func<EntityUid, bool>? filter = null)
+    {
+        var stations = new ValueList<EntityUid>(Count<StationEventEligibleComponent>());
+
+        filter ??= _ => true;
+        var query = AllEntityQuery<StationEventEligibleComponent>();
+
+        while (query.MoveNext(out var uid, out _))
+        {
+            if (!filter(uid))
+                continue;
+
+            stations.Add(uid);
+        }
+
+        if (stations.Count == 0)
+        {
+            station = null;
+            return false;
+        }
+
+        // TODO: Engine PR.
+        station = stations[RobustRandom.Next(stations.Count)];
+        return true;
+    }
```

Comment:
I made an engine pr without a filter but this should also really be an engine pr.

Reply:
this is just ported from `StationEventSystem`, i didnt add this

Reply:
also stations arent engine?